src/rgw/rgw_sal.h

```diff
@@ -1498,20 +1499,29 @@ class Zone {
 };
 
 /**
- * @brief Abstraction of a manager for Lua scripts
+ * @brief Abstraction of a manager for Lua scripts and packages
  *
- * RGW can load and process Lua scripts.  This will handle loading/storing scripts.
+ * RGW can load and process Lua scripts.  This will handle loading/storing
+ * scripts; adding, deleting, listing, and installing packages.
  */
-class LuaScriptManager {
+class LuaManager {
 public:
-  virtual ~LuaScriptManager() = default;
+  virtual ~LuaManager() = default;
 
   /** Get a script named with the given key from the backing store */
-  virtual int get(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, std::string& script) = 0;
+  virtual int get_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, std::string& script) = 0;
   /** Put a script named with the given key to the backing store */
-  virtual int put(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, const std::string& script) = 0;
+  virtual int put_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, const std::string& script) = 0;
   /** Delete a script named with the given key from the backing store */
-  virtual int del(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key) = 0;
+  virtual int del_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key) = 0;
+  /** Add a lua package */
+  virtual int add_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name, bool allow_compilation) { return -ENOENT; }
+  /** Remove a lua package */
+  virtual int remove_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name) { return -ENOENT; }
+  /** List lua packages */
+  virtual int list_packages(const DoutPrefixProvider* dpp, optional_yield y, rgw::lua::packages_t& packages) { return -ENOENT; }
+  /** Install lua packages */
+  virtual int install_packages(const DoutPrefixProvider* dpp, optional_yield y, rgw::lua::packages_t& failed_packages, std::string& output) { return -ENOENT; }
```

Comment:
as mentioned here: https://github.com/ceph/ceph/pull/46980/files#r917355672

we can have the default implementation

Reply:
but in rgw_sal_store.h

Reply:
"sal" is the base interface for anything.  "sal_store" is the base implementation for stores (actual backing stores, living at the bottom of a stack). "sal_filter" is the base implementation (still under development) for filters, which live somewhere in the middle of the stack.